test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java

```diff
@@ -0,0 +1,306 @@
+     * being thrown when the method is called.
+     *
+     * @param method  the method for which to invoke the handler
+     * @param handler the handler
+     *
+     * @return this object
+     * @throws IllegalArgumentException if the method is not declared in {@code JavaFileManager}
+     * @throws IllegalStateException if a handler is already registered for this method
+     */
+    public TestJavaFileManagerBuilder handle(Method method,
+                                             BiFunction<JavaFileManager, Object[], Throwable> handler) {
+        if (!JavaFileManager.class.isAssignableFrom(method.getDeclaringClass())) {
+            throw new IllegalArgumentException(("not a method on JavaFileManager: " + method));
+        }
+
+        var prev = fileManagerHandlers.put(method, handler);
+        if (prev != null) {
+            throw new IllegalStateException("handler already registered for method " + method);
+        }
+        return this;
+    }
+
+    /**
+     * Provides a function to be called when a given file object method is called,
+     * for file objects that match a given predicate.
+     * The function should either return an exception to be thrown, or {@code null}
+     * to indicate that no exception should be thrown.
+     *
+     * It is an error for the function to return a checked exception that is not
+     * declared by the method. This error will result in {@link UndeclaredThrowableException}
+     * being thrown when the method is called.
+     *
+     * @apiNote Examples of predicates include:
+     * <ul>
+     * <li>using {@code .equals} or {@link JavaFileObject#isNameCompatible(String, JavaFileObject.Kind)}
+     *     to match a specific file object,
+     * <li>using string or regular expression operations on the name or URI of the file object,
+     * <li>using {@code Path} operations on the file object's {@link StandardJavaFileManager#asPath(FileObject) path}.
+     * </ul>
+     *
+     * @param filter  the predicate used to identify file objects for this handler
+     * @param method  the method for which to invoke the handler
+     * @param handler the handler
+     *
+     * @return this object
+     * @throws IllegalArgumentException if the method is not declared in a class that is assignable
+     *          to {@code FileObject}
+     * @throws IllegalStateException if a handler is already registered for this predicate and method
+     */
+    public TestJavaFileManagerBuilder handle(Predicate<JavaFileObject> filter,
+                                             Method method,
+                                             BiFunction<JavaFileObject, Object[], Throwable> handler) {
+        if (!FileObject.class.isAssignableFrom(method.getDeclaringClass())) {
+            throw new IllegalArgumentException(("not a method on FileObject: " + method));
+        }
+
+        var map = fileObjectHandlers.computeIfAbsent(filter, p_ -> new HashMap<>());
+        var prev = map.put(method, handler);
```

Comment:
I've modified the API so that it is no longer necessary to use the predicate as a key.

Instead, the map is no longer built incrementally by a series of `handle` calls;  the signature of the `handle` call is modified to take the predicate and a user-provided map.  It is expected that maps will be small (typically one element) and easy to create with `Map.of`.    While this specifically applies to the handling of file objects, for consistency a similar change is made to the API to register handlers for file manager calls.

Also of note, because this removes the need to create the maps within the builder class, the other issue about using `putIfAbsent` instead of `put` is now moot.